### ISSUE TITLE
Use `objc_msgSend_stret` and `objc_msgSend_fpret` instead of `objc_msgSend` for signature that require that

### ISSUE THREAD

**Issue by dnys1** [dart-lang/ffigen#608]
Repro: https://github.com/dnys1/ffigen_crash

We received a [crash report](https://github.com/aws-amplify/amplify-flutter/issues/3630) which I traced back to this code:

```dart
final screen = NSScreen.getMainScreen(lib);
screen!.visibleFrame;
//           ^ crashes on access to visibleFrame
```

The issue is not reproducible on Silicon macOS and was reported from a user with an Intel machine. I was able to reproduce it by running the repro app above under Rosetta, which I learned can be enabled in Xcode as a run destination under **Product** > **Destination** > **Destination Architectures** > **Show Both**.

Both myself and the user are running on Flutter 3.13 and Xcode `14.3.1`. This issue is reproducible on Flutter 3.10 and Flutter beta as well.

Thanks for any help!

**Comment by liamappelbe** [comment]
I tried a few other methods on this class, and it looks like this crash happens for methods that try to return a struct by value. The methods that return an object by pointer or a primitive by value work fine. So my first thought was that returning structs by value might be broken on this architecture, but the Dart SDK test `tests/ffi/function_structs_by_value_generated_ret_test.dart` works fine.

The only thing that might be different about the ObjC case is that all these method invocations work through the same `objc_msgSend` function. We load it with many different signatures, one for each method signature. So we're relying on the assembly language that implements `objc_msgSend` to obey the C calling convention for that signature. I was able to repro this crash in pure C (see below), so that means it's not a problem on the Dart side.

I have 2 guesses:

1) The `CGRect` struct might have changed between versions of the framework. If the struct has this 4 doubles shape in the header, but in the actual framework it has a different shape, then that would likely seg fault.

2) `objc_msgSend` might not be exactly following the C calling convention. From the little I've looked into Dart's native wrapper IL, the calling convention looks really complicated and architecture specific, and the struct handling especially so. If there are any small differences between the way `objc_msgSend` returns structs by value, and C or Dart expects to receive the result, that would likely cause a seg fault. If this is the case, I'm not sure what the fix should be. I don't want to have to special case `objc_msgSend`.

Next steps should be to figure out which of those 2 options is the case.

```c
#include <stdio.h>

void* sel_registerName(const char* name);
void* objc_getClass(const char* name);
void objc_msgSend();

typedef struct CGPoint {
  double x;
  double y;
} CGPoint;

typedef struct CGSize {
  double width;
  double height;
} CGSize;

typedef struct CGRect {
  CGPoint origin;
  CGSize size;
} CGRect;

int main() {
  void* class_NSScreen = objc_getClass("NSScreen");
  printf("NSScreen: %p\n", class_NSScreen);

  void* sel_mainScreen = sel_registerName("mainScreen");
  void* sel_maximumFramesPerSecond = sel_registerName("maximumFramesPerSecond");
  void* sel_visibleFrame = sel_registerName("visibleFrame");

  void*(*send_mainScreen)(void*,void*) = (void*(*)(void*,void*))&objc_msgSend;
  void* mainScreen = send_mainScreen(class_NSScreen, sel_mainScreen);
  printf("mainScreen: %p\n", mainScreen);

  long(*send_maximumFramesPerSecond)(void*,void*) =
      (long(*)(void*,void*))&objc_msgSend;
  long maximumFramesPerSecond =
      send_maximumFramesPerSecond(mainScreen, sel_maximumFramesPerSecond);
  printf("maximumFramesPerSecond: %ld\n", maximumFramesPerSecond);

  CGRect(*send_visibleFrame)(void*,void*) =
      (CGRect(*)(void*,void*))&objc_msgSend;
  CGRect visibleFrame = send_visibleFrame(mainScreen, sel_visibleFrame);  // SEG_FAULT here
  printf("visibleFrame: {(%lf, %lf), (%lf, %lf)}\n", visibleFrame.origin.x,
      visibleFrame.origin.y, visibleFrame.size.width, visibleFrame.size.height);
}
```

`clang test.c -framework Foundation -framework AppKit -o test && test`

**Comment by dcharkes** [comment]
**TL;DR: Use `objc_msgSend_stret` for calling methods that return structs on x64 (not on arm64).**

```c
#include <CoreGraphics/CGBase.h>
#include <objc/message.h>
#include <stdio.h>

int main() {
  void *class_NSScreen = objc_getClass("NSScreen");
  printf("NSScreen: %p\n", class_NSScreen);

  void *sel_mainScreen = sel_registerName("mainScreen");
  void *sel_maximumFramesPerSecond = sel_registerName("maximumFramesPerSecond");
  void *sel_visibleFrame = sel_registerName("visibleFrame");

  void *(*send_mainScreen)(void *, void *) =
      (void *(*)(void *, void *)) & objc_msgSend;
  void *mainScreen = send_mainScreen(class_NSScreen, sel_mainScreen);
  printf("mainScreen: %p\n", mainScreen);

  long (*send_maximumFramesPerSecond)(void *, void *) =
      (long (*)(void *, void *)) & objc_msgSend;
  long maximumFramesPerSecond =
      send_maximumFramesPerSecond(mainScreen, sel_maximumFramesPerSecond);
  printf("maximumFramesPerSecond: %ld\n", maximumFramesPerSecond);

  CGRect (*send_visibleFrame)(void *, void *) =
      (CGRect(*)(void *, void *)) & objc_msgSend_stret;
  CGRect visibleFrame =
      send_visibleFrame(mainScreen, sel_visibleFrame); // SEG_FAULT here
  printf("visibleFrame: {(%lf, %lf), (%lf, %lf)}\n", visibleFrame.origin.x,
         visibleFrame.origin.y, visibleFrame.size.width,
         visibleFrame.size.height);
}
```

```
$ arch -arch x86_64 clang test.c -framework Foundation -framework AppKit -o test
$ ./test 
NSScreen: 0x7ff8576a0178
mainScreen: 0x600000a29680
maximumFramesPerSecond: 120
visibleFrame: {(0.000000, 87.000000), (1728.000000, 992.000000)}
```


--------------

Long road of discovery:

```
$ clang test.c -framework Foundation -framework AppKit
$ ./test 
NSScreen: 0x1fbee9368
mainScreen: 0x600003949380
maximumFramesPerSecond: 120
visibleFrame: {(0.000000, 87.000000), (1728.000000, 992.000000)}

$ arch -arch x86_64 clang test.c -framework Foundation -framework AppKit -o test
$ ./test 
NSScreen: 0x7ff8576a0178
mainScreen: 0x6000020d9620
maximumFramesPerSecond: 120
Segmentation fault: 11
```

I can repro this in x64 mode on M1.

> The CGRect struct might have changed between versions of the framework

If I change the structs to be imported from the header, the issue persistents on x64.

```
#include <CoreGraphics/CGBase.h>
#include <objc/message.h>
#include <stdio.h>
```

So, unfortunately, it's unlikely that the layout of the struct changed.

For reference, the headers are in found inside Xcode: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/System/Library/Frameworks/CoreFoundation.framework/Versions/A/Headers/CFCGTypes.h

> 2\. `objc_msgSend` might not be exactly following the C calling convention.

Firing the test executable up in lldb:

```
-exec bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x18)
  * frame #0: 0x00007ff813c0a75d libobjc.A.dylib`objc_msgSend + 29
    frame #1: 0x0000000100003ed5 test`main + 213
    frame #2: 0x000000020001241f dyld`start + 1903

-exec x/20i $pc-29
    0x7ff813c0a740: 48 85 ff                       testq  %rdi, %rdi
    0x7ff813c0a743: 2e 74 72                       je     0x7ff813c0a7b8            ; <+120>
    0x7ff813c0a746: 40 f6 c7 01                    testb  $0x1, %dil
    0x7ff813c0a74a: 2e 75 76                       jne    0x7ff813c0a7c3            ; <+131>
    0x7ff813c0a74d: 49 ba f8 ff ff ff ff 7f 00 00  movabsq $0x7ffffffffff8, %r10     ; imm = 0x7FFFFFFFFFF8 
    0x7ff813c0a757: 4c 23 17                       andq   (%rdi), %r10
    0x7ff813c0a75a: 49 89 f3                       movq   %rsi, %r11
->  0x7ff813c0a75d: 45 23 5a 18                    andl   0x18(%r10), %r11d
    0x7ff813c0a761: 49 c1 e3 04                    shlq   $0x4, %r11

-exec register read
General Purpose Registers:
       rax = 0x00007ff813c0a740  libobjc.A.dylib`objc_msgSend
       rbx = 0x0000000100003e00  test`main
       rcx = 0xafb80379ba400059
       rdx = 0x00007ff833873b7a  
       rdi = 0x000000030410b088
       rsi = 0x0000000108c382e0
       rbp = 0x000000030410b0f0
       rsp = 0x000000030410b078
        r8 = 0x00007ff857680f88  __sFX + 248
        r9 = 0x0000000000000000
       r10 = 0x0000000000000000      // This is indeed 0, leading to a bad memory access.
       r11 = 0x0000000108c382e0
       r12 = 0x000000030410b2b8
       r13 = 0x000000030410b2f0
       r14 = 0x000000030410b300
       r15 = 0x000000030410b180
       rip = 0x00007ff813c0a75d  libobjc.A.dylib`objc_msgSend + 29
// ...
```

We're crashing quite early inside `objc_msgSend`. It seems that the assembly code is handwritten: https://www.mikeash.com/pyblog/friday-qa-2012-11-16-lets-build-objc_msgsend.html Since we're crashing so early, it might be the lookup or the cache of the method in the class.

While reading about msgsend I stumbled upon:

> Another case where just calling objc_msgSend straight wouldn't work is a method that returns an NSRect, say, because objc_msgSend is not used in that case, objc_msgSend_stret is. In the underlying C function for a method that returns an NSRect, the first argument is actually a pointer to an out value NSRect, and the function itself actually returns void. You must match this convention when calling because it's what the called method will assume. Further, the circumstances in which objc_msgSend_stret is used differ between architectures. There is also an objc_msgSend_fpret, which should be used for methods that return certain floating point types on certain architectures.

https://stackoverflow.com/questions/2573805/using-objc-msgsend-to-call-a-objective-c-function-with-named-arguments

This solves the issue.

**Comment by dnys1** [comment]
Awesome investigation. Thanks for diving into this!

**Comment by liamappelbe** [comment]
> TL;DR: Use `objc_msgSend_stret` for calling methods that return structs on x64 (not on arm64).

Thanks for the help @dcharkes. Reading [the docs](https://developer.apple.com/documentation/objectivec/1456730-objc_msgsend_stret) for the stret and fpret variations, it sounds like we should be using these on all platforms. What makes you say not on arm64? My understanding is that while they're not *necessary* on arm64, you can still use them. I'd rather not generate architecture specific code if I don't have to.

**Comment by dcharkes** [comment]
> Thanks for the help @dcharkes. Reading [the docs](https://developer.apple.com/documentation/objectivec/1456730-objc_msgsend_stret) for the stret and fpret variations, it sounds like we should be using these on all platforms. What makes you say not on arm64? My understanding is that while they're not _necessary_ on arm64, you can still use them. I'd rather not generate architecture specific code if I don't have to.

Ah yes, cleaner to use on all archs. (I'm not sure what made me say to not use it, I don't see anything in the related links or my notes.)
